Editorial pass of C# tutorial (#17704)
microsoft/onnxruntime: docs/tutorials/csharp/basic_csharp.md

# C# Tutorial: Basic

Learn how to get started with inference with the C# API. Run an ONNX model for a given input data. The model is typically trained using any of the well-known training frameworks and then exported into the ONNX format.

## OrtValue API

The new `OrtValue` based API is the recommended approach. The `OrtValue` API generates less garbage and is more performant. Some scenarios show 4x performance improvement over the previous API and significantly less garbage.

OrtValue is a universal container that can hold different ONNX types, such as tensors, maps, and sequences. It always existed in the onnxruntime library, but was not exposed in the C# API.

The `OrtValue` based API provides uniform access to data via `ReadOnlySpan<T>` and `Span<T>` structures, regardless of its location, managed or unmanaged.

Note, that the following classes `NamedOnnxValue`, `DisposableNamedOnnxValue`, `FixedBufferOnnxValue` will be deprecated in the future. They are not recommended for new code.

## Data shape

> `DenseTensor` class can be used for multi-dimensional access to the data since the new `Span` based API features only a 1-D index. However, some reported a slow performance when using `DenseTensor` class multi-dimensional access. One can then create an OrtValue on top of the tensors data. 

Reviewer: ```suggestion
`DenseTensor` class can be used for multi-dimensional access to the data since the new `Span` based API features only a 1-D index. However, some reported a slow performance when using `DenseTensor` class multi-dimensional access. One can then create an OrtValue on top of the DenseTensor
```suggestion
`DenseTensor` class can be used for multi-dimensional access to the data since the new `Span` based API features only a 1-D index. However, some reported a slow performance when using `DenseTensor` class multi-dimensional access. One can then create an OrtValue on top of the tensors data. 
``` data. 
```

`ShapeUtils` class provides some help to deal with multi-dimensional indices for OrtValues.

If output shapes are known, one can pre-allocate `OrtValue` on top of the managed or unmanaged allocations and supply those OrtValues to be used as outputs. Due to this fact, the need for `IOBinding` is greatly diminished.

## Data types

`OrtValues` can be created directly on top of the managed `unmanaged` [struct based blittable types](https://learn.microsoft.com/en-us/dotnet/framework/interop/blittable-and-non-blittable-types) arrays. The onnxruntime C# API allows use of managed buffers for input or output.

String data is represented as UTF-16 string objects in C#. It will still need to be copied and converted to UTF-8 to the native memory. However, that conversion is now more optimized and is done in a single pass without intermediate byte arrays.

The same applies to string `OrtValue` tensors returned as outputs. Character based API now operates on `Span<char>`,`ReadOnlySpan<char>`, and `ReadOnlyMemory<char>` objects. This adds flexibility to the API and allows to avoid unnecessary copies.

## Data life-cycle

> Except some of the above deprecated API classes, nearly all of C# API classes are `IDisposable`.

Reviewer: ```suggestion
Except for some of the above deprecated API classes, nearly all of C# API classes are `IDisposable`.
```

# C# Tutorial: Basic

> Learn how to get started with inference with the C# API. Run an ONNX model for a given input data. The model is typically trained using any of the well-known training frameworks and then exported into the ONNX format.

Reviewer 2: I think we can simplify this section to just a single sentence: "Learn how to get started with inferencing ONNX models using the ONNX Runtime C# API"

Do not hold your breath for finalizers which are not guaranteed to ever run, and if they do, they do it when it is too late.

This includes `SessionOptions`, `RunOptions`, `InferenceSession`, `OrtValue`. Run() calls return `IDisposableCollection` that allows to dispose all of the containing objects in one statement or `using`. This is because these objects own native resources, often a native object.

Not disposing `OrtValue` that was created on top of the managed buffer would result in
that buffer pinned in memory indefinitely. Such a buffer can not be garbage collected or moved in memory.

`OrtValue`s that were created on top of the native onnxruntime memory should also be disposed of promptly. Otherwise, the native memory will not be deallocated. OrtValues returned by `Run()` usually hold native memory.

GC can not operate on native memory or any other native resources.

The `using` statement or a block is a convenient way to ensure that the objects are disposed.

> `InferenceSession` can be a long lived object and a member of another class. It eventually must also need to be disposed. This means, the containing class also would have to be made disposable to achieve this.

Reviewer 2: extra" need"

> Once a session is created, you can execute queries using the `Run` method of the  `InferenceSession` object.

Reviewer: ```suggestion
Once a session is created, you can run inference using the `Run` method of the  `InferenceSession` object.
```

{
    strTensor.StringTensorSetElementAt(strs[i].AsSpan(), i);
}
```

## More examples

* [Stable Diffusion](stable-diffusion-csharp.md)

> * [BERT NLP example](bert-nlp-csharp-console-app.md)

Reviewer 2: Can exclude "example" since it's already in the header for the section
